File: CSharp/Oomph.Data/UF09Test/UFs/v301/Int3/ABC214_D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Oomph.Data.UF09Lib.UFs.v301;

namespace UF09Test.UFs.v301.Int3
{
	// Test: https://atcoder.jp/contests/abc214/tasks/abc214_d
	class ABC214_D
	{
		static int[] Read() => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
		static (int, int, int w) Read3() { var a = Read(); return (a[0], a[1], a[2]); }
		static void Main() => Console.WriteLine(Solve());
		static object Solve()
		{
			var n = int.Parse(Console.ReadLine());
			var es = Array.ConvertAll(new bool[n - 1], _ => Read3());

			var r = 0L;
			var uf = new UnionFind(n + 1);

			foreach (var (u, v, w) in es.OrderBy(e => e.w))
			{
				r += (long)w * uf.Find(u).Size * uf.Find(v).Size;
				uf.Union(u, v);
			}
			return r;
		}
	}
}

File: Python/oomph-data-ws/uf_309_test/ufs/v301/L1/ABC214_D.py
# https://atcoder.jp/contests/abc214/tasks/abc214_d

from .....uf_309_lib.ufs.int_omega.unionfind_301 import UnionFind

n = int(input())
es = [list(map(int, input().split())) for j in range(n - 1)]

r = 0
uf = UnionFind(n + 1)

es.sort(key=lambda e: e[2])
for u, v, w in es:
    r += uf.find(u).size * uf.find(v).size * w
    uf.union(u, v)

print(r)
